analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LoopsAndLinq.CSharp11.cs

```diff
@@ -1,37 +1,99 @@
+        {
+            if (array is [1, ..] local)
+            {
+                Console.WriteLine("Pattern match successful");
             }
+
+        }
+    }
+}
+
+// https://github.com/SonarSource/sonar-dotnet/issues/7730
+class Repro_7730
+{
+    void SpansAndLogicalPatterns(Span<char> s, ReadOnlySpan<char> ros)
+    {
+        foreach (var c in s)                   // Noncompliant, FP: iterable but not enumerable, nor queriable
+            if (c is ' ') { }                  // Secondary
+        foreach (var c in s)                   // Noncompliant, FP
+            if (c is not ' ') { }              // Secondary
+        foreach (var c in s)                   // Noncompliant, FP
+            if (c is ' ' or '\n' or '\r') { }  // Secondary
+
+        foreach (var c in ros)                 // Noncompliant, FP
+            if (c is ' ') { }                  // Secondary
+        foreach (var c in ros)                 // Noncompliant, FP
+            if (c is not ' ') { }              // Secondary
+        foreach (var c in ros)                 // Noncompliant, FP
+            if (c is ' ' or '\n' or '\r') { }  // Secondary
+    }
+
+    void SpansAndLogicalOperators(Span<char> s, ReadOnlySpan<char> ros)
+    {
+        foreach (var c in s)                             // Compliant: iterable but not enumerable, nor queriable
+            if (c == ' ') { }
+        foreach (var c in s)                             // Compliant
+            if (c != ' ') { }
+        foreach (var c in s)                             // Compliant
+            if (c == ' ' || c == '\n' || c == '\r') { }
+
+        foreach (var c in ros)                           // Compliant
+            if (c == ' ') { }
+        foreach (var c in ros)                           // Compliant
+            if (c != ' ') { }
+        foreach (var c in ros)                           // Compliant
+            if (c == ' ' || c == '\n' || c == '\r') { }
+    }
+
+    void IterableNotEnumerableAndLogicalPatterns(IterableNotEnumerable s)
+    {
+        foreach (var c in s)                   // Noncompliant, FP: iterable but not enumerable, nor queriable
+            if (c is ' ') { }                  // Secondary
+        foreach (var c in s)                   // Noncompliant, FP
+            if (c is not ' ') { }              // Secondary
+        foreach (var c in s)                   // Noncompliant, FP
+            if (c is ' ' or '\n' or '\r') { }  // Secondary
+    }
+
+    void ListAndLogicalOperators(List<char> s)
+    {
+        foreach (var c in s)                   // FN, equivalent to c is ' '
+            if (c == ' ') { }
+        foreach (var c in s)                   // FN, equivalent to c is not ' '
+            if (c != ' ') { }
+    }
+
+    class IterableNotEnumerable
+    {
+        public IEnumerator<char> GetEnumerator()
+        {
+            yield return 'a';
+            yield return 'b';
         }
     }
```

Comment:
Can you add also a test iterating a custom class that is inheriting a list:
```cs
var myList = new MyList();
foreach(var x in myList) { }
public class MyList : List<int>{}
```

and does this rule raises with methods/delegates?
```cs
public void ThroughLambda(Func<ImmutableList<int>> lambda)
{
    foreach(var x in MyMethod()) { }
    foreach(var x in lambda()) { }

    List<int> MyMethod() => new();
}
```

Reply:
Added tests iterating on custom lists and also custom `IEnumerable<T>`.

Yes, the rule behaves correctly in the presence of methods, delegates, and local functions.
I have added test cases showing that.